### PR TITLE
FIX: Bindings pasted to Action appear at top of the list (ISX-1899)

### PR DESCRIPTION
**Pull request by timkeo**
### Description

Fixes a couple copy/paste issues in the UITK Editor in which pasted bindings appeared at the top of the list instead of the bottom.

### Changes made

Modified the logic within `PasteBindingsFromClipboard()` and `PasteBindingOrComposite()` to update the "index" of pasted items so it appears at the bottom of their respective sets when their "root" node is selected:

- Bindings pasted to an Action are inserted at the end of the Action's current Binding list
- CompositePart Bindings are inserted at the end of the Component Binding's current set

When a Binding or CompositePart is selected during paste, the pasted item will still appear immediately below the selected item.

### Notes

I was only able to do a little of bit of testing myself, and would've like to do more before opening the PR but ran out of time. I figured it's better to just get the PR open and deal with any issues next week.

Having automated tests to cover these scenarios would be really good, but that'll take some time/effort, which I didn't think is appropriate at this point.

### Checklist

Before review:

- [ ] Changelog entry added.
    - Explains the change in `Changed`, `Fixed`, `Added` sections.
    - For API change contains an example snippet and/or migration example.
    - FogBugz ticket attached, example `([case %number%](https://issuetracker.unity3d.com/issues/...))`.
    - FogBugz is marked as "Resolved" with *next* release version correctly set.
- [ ] Tests added/changed, if applicable.
    - Functional tests `Area_CanDoX`, `Area_CanDoX_EvenIfYIsTheCase`, `Area_WhenIDoX_AndYHappens_ThisIsTheResult`.
    - Performance tests.
    - Integration tests.
- [ ] Docs for new/changed API's.
    - Xmldoc cross references are set correctly.
    - Added explanation how the API works.
    - Usage code examples added.
    - The manual is updated, if needed.

During merge:

- [ ] Commit message for squash-merge is prefixed with one of the list:
    - `NEW: ___`.
    - `FIX: ___`.
    - `DOCS: ___`.
    - `CHANGE: ___`.
    - `RELEASE: 1.1.0-preview.3`.

### CODE REVIEW

File: Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/CopyPasteHelper.cs
Comment (ritamerkl) [inline Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/CopyPasteHelper.cs]: the if/else have identical content. maybe declare boolean variables before and check them into one combined if statement?

File: Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/CopyPasteHelper.cs
Comment (ritamerkl) [inline Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/CopyPasteHelper.cs]: The new guidelines recommend to not use Linq anymore. I know it is on various places in our code already, but I think it makes sense to try not using it in the new/changed code.

**Review by ritamerkl — COMMENTED** [review]
LGTM! I think there is a changelog entry missing.

File: Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/CopyPasteHelper.cs
Comment (ekcoh) [reply Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/CopyPasteHelper.cs]: Agreed, lets try to avoid introducing more.

File: Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/CopyPasteHelper.cs
Comment (ekcoh) [inline Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/CopyPasteHelper.cs:229]: Minor: Would suggest not mixing brackets and non brackets, i.e. use brackets on the else.

**Review by ekcoh — APPROVED** [review]
Looks good, some minor comments. Comments from @ritamerkl should be addressed but given they are this looks good from my perspective.

**Review by Pauliusd01 — CHANGES_REQUESTED** [review]
This breaks pasting composite bindings into non composite bindings checks. Steps to repro: Create a composite binding action -> Copy or Cut and paste it to a non composite binding -> errors thrown (No repro in Develop)

https://github.com/Unity-Technologies/InputSystem/assets/54306142/11d74de8-94bb-4f45-af2f-77368b94c81f

File: Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/CopyPasteHelper.cs
Comment (timkeo) [reply Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/CopyPasteHelper.cs]: The "Select()" call from `bindingsForAction.Select`, which I copied over from `/UITKAssetEditor/Views/Selectors.cs` is what's using it.

I'll restore the function `Selectors.GetLastBindingIndexForSelectedAction()`, which contained this logic (Selectors.cs already uses LINQ) and call into that from here.

File: Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/CopyPasteHelper.cs
Comment (timkeo) [reply Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/CopyPasteHelper.cs]: The logic was kinda confusing for me, and so I did it this way because I found it easier to reason through, but I do agree it's not optimal. I'll see if can rework it so it's not duplicating checks but still not too confusing.

Comment (ritamerkl) [reply Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/CopyPasteHelper.cs]: How about if you take the content of the if statements into two booleans with comments?

Comment (timkeo) [reply Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/CopyPasteHelper.cs]: Yeah that's a good suggestion, and I think I'll go that way.

Comment (timkeo) [reply Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/CopyPasteHelper.cs]: Reworked the conditions to remove duplicated line, and also fix regression pasting Composite: 630fe8d551ecfe45a0245c77583f76b8faf1f285

File: Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/CopyPasteHelper.cs
Comment (timkeo) [reply Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/CopyPasteHelper.cs]: Moved the code using LINQ back into Selectors.cs with: 630fe8d551ecfe45a0245c77583f76b8faf1f285

File: Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/CopyPasteHelper.cs
Comment (timkeo) [reply Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/CopyPasteHelper.cs:229]: Fixed with 630fe8d551ecfe45a0245c77583f76b8faf1f285.

**Review by ritamerkl — APPROVED** [review]
As far as I see still no changelog or do we skip that for this fix? 
Looks good for me, would be fine to merge after Paulius approved as well.

**Comment by Pauliusd01** [comment]
Pasting composite bindings into non composite actions is now correctly blocked but it also clears the clipboard when trying to paste incorrectly. For example: You cut a composite binding and try to paste it into a non composite action -> nothing happens -> you try again on a composite action -> still nothing happens (On Develop it pastes correctly in this step)

**Review by Pauliusd01 — CHANGES_REQUESTED** [review]
forgot to update status

**Comment by timkeo** [comment]
> As far as I see still no changelog or do we skip that for this fix? Looks good for me, would be fine to merge after Paulius approved as well.

Ah you're right, I completely missed that!
I guess I see this fix as a Multi-Copy-Paste feature detail and probably not something that needs to be called out.

**Comment by timkeo** [comment]
> Pasting composite bindings into non composite actions is now correctly blocked but it also clears the clipboard when trying to paste incorrectly. For example: You cut a composite binding and try to paste it into a non composite action -> nothing happens -> you try again on a composite action -> still nothing happens (On Develop it pastes correctly in this step)

I was unable to repro this, but I did find and fix the other issue you came across (errors logged to the console in when copy/pasting CompositeParts). This was due to a flaw in `GetSelectedBindingIndexAfterCompositeBindings()`, which wasn't checking for `index` going beyond the array size. This issue only occurs when the Composite binding you're trying to paste into is last in the Actions.

The specific issue you pointed out didn't occur in develop because we weren't trying to paste at the end of the Composite list before. However, I did some additional testing and found another case where this error occurs (in Develop), and verified the change fixes those as well.

In general, I see a lot of places where we call `GetArrayElementAtIndex()` for a Bindings property without checking the `index` against `arraySize`. I think this is generally fine if/when we expect `bindingIndex` to be valid, but we should be testing cut/copy/paste/delete scenarios against Bindings (regular and Composite) at the bottom of the Actions list.

**Comment by Pauliusd01** [comment]
Cutting and pasting bindings into an empty action throws nullrefs

https://github.com/Unity-Technologies/InputSystem/assets/54306142/8ce93b01-bd44-4b64-a0b7-ccf975455e06

Trying to cut a 2nd time after pasting bindings with composites somehow only cuts the action properties and throws nullrefs 

https://github.com/Unity-Technologies/InputSystem/assets/54306142/f2b9aead-6e3d-41a0-b37c-33964e6de7a3

**Review by Pauliusd01 — CHANGES_REQUESTED** [review]
Updating status - waiting for fixes

**Comment by timkeo** [comment]
> Cutting and pasting bindings into an empty action throws nullrefs
> Unity_2024-02-28_12-54-31.mp4
> 
> Trying to cut a 2nd time after pasting bindings with composites somehow only cuts the action properties and throws nullrefs
> Unity_2024-02-28_12-53-26.mp4

Both of these issues seems to stem from the same problem: we don't properly set the `selectedBindingIndex` value after cutting a Binding. This _should_ be solved with @ritamerkl [Cut/Paste/Delete fix PR](https://github.com/Unity-Technologies/InputSystem/pull/1858), since Cut doesn't delete the item immediately and avoids us getting into this bad state. 

I tested my changes with Rita's branch and wasn't able to repro either of these issues, However, I decided to fix the 1st issue (pasting to an empty Action) by adding another empty Bindings Array check. Although it should be unnecessary with Rita's work it's a simple, safe change. The 2nd issue is much more complex and can't be easily fixed without duplicating/conflicting with Rita's changes.

I propose opening a ticket to re-investigate these issue (and `selectedBindingIndex` in general) after both our PRs are merged and address any remaining problems then.

**Comment by Pauliusd01** [comment]
> > Cutting and pasting bindings into an empty action throws nullrefs
> > Unity_2024-02-28_12-54-31.mp4
> > Trying to cut a 2nd time after pasting bindings with composites somehow only cuts the action properties and throws nullrefs
> > Unity_2024-02-28_12-53-26.mp4
> 
> Both of these issues seems to stem from the same problem: we don't properly set the `selectedBindingIndex` value after cutting a Binding. This _should_ be solved with @ritamerkl [Cut/Paste/Delete fix PR](https://github.com/Unity-Technologies/InputSystem/pull/1858), since Cut doesn't delete the item immediately and avoids us getting into this bad state.
> 
> I tested my changes with Rita's branch and wasn't able to repro either of these issues, However, I decided to fix the 1st issue (pasting to an empty Action) by adding another empty Bindings Array check. Although it should be unnecessary with Rita's work it's a simple, safe change. The 2nd issue is much more complex and can't be easily fixed without duplicating/conflicting with Rita's changes.
> 
> I propose opening a ticket to re-investigate these issue (and `selectedBindingIndex` in general) after both our PRs are merged and address any remaining problems then.

Thank you Tim, It's a bit confusing to review when there's multiple ongoing PRs on the same thing. I'll approve this, we will have to QA the full release anyway and hopefully any missed bugs should spring up then

**Review by Pauliusd01 — APPROVED** [review]

**Pull request merged**
